Unsafe within an expression

I tried this code:

```rust
pub fn test() -> i32 {
    let a = unsafe { 123 };
    a
}

pub fn main() {
    let a = test();
}
```

I expected to see this happen: *explanation*

Compiled sucsessfully 

Instead, this happened: *explanation*

```
ust1: note: Preparing to parse files. 
rust1: note: Attempting to parse file: test.rs
rust1: note: null denotation is null; returning null for parse_expr
test.rs:14:13: error: found unexpected token ‘unsafe’ in null denotation
   14 |     let a = unsafe { 123 };
      |             ^
test.rs:14:20: error: failed to parse expression in let statement
   14 |     let a = unsafe { 123 };
      |                    ^
test.rs:14:5: error: failed to parse statement or expression without block in block expression
   14 |     let a = unsafe { 123 };
      |     ^
test.rs:15:5: error: unrecognised token ‘identifier’ for start of item
   15 |     a
      |     ^
test.rs:15:5: error: failed to parse item in crate
test.rs:16:1: error: attempted to get string for ‘}’, which has no string. returning empty string instead
   16 | }
      | ^
test.rs:18:1: error: attempted to get string for ‘pub’, which has no string. returning empty string instead
   18 | pub fn main() {
      | ^
test.rs:18:5: error: attempted to get string for ‘fn’, which has no string. returning empty string instead
   18 | pub fn main() {
      |     ^
test.rs:18:12: error: attempted to get string for ‘(’, which has no string. returning empty string instead
   18 | pub fn main() {
      |            ^
test.rs:18:13: error: attempted to get string for ‘)’, which has no string. returning empty string instead
   18 | pub fn main() {
      |             ^
test.rs:18:15: error: attempted to get string for ‘{’, which has no string. returning empty string instead
   18 | pub fn main() {
      |               ^
test.rs:19:5: error: attempted to get string for ‘let’, which has no string. returning empty string instead
   19 |     let a = test();
      |     ^
test.rs:19:11: error: attempted to get string for ‘=’, which has no string. returning empty string instead
   19 |     let a = test();
      |           ^
test.rs:19:17: error: attempted to get string for ‘(’, which has no string. returning empty string instead
   19 |     let a = test();
      |                 ^
test.rs:19:18: error: attempted to get string for ‘)’, which has no string. returning empty string instead
   19 |     let a = test();
      |                  ^
test.rs:19:19: error: attempted to get string for ‘;’, which has no string. returning empty string instead
   19 |     let a = test();
      |                   ^
test.rs:20:1: error: attempted to get string for ‘}’, which has no string. returning empty string instead
   20 | }
      | ^
```

### Meta

- What version of Rust GCC: 6dfde8225f4184552aca9c614894ac13144c2298

--- a/gcc/testsuite/rust/compile/torture/unsafe3.rs
+++ b/gcc/testsuite/rust/compile/torture/unsafe3.rs
@@ -1,0 +1,9 @@
+pub fn test() -> i32 {
+    let a = unsafe { 123 };
+    a
+}
+
+pub fn main() {
+    let a = test();
+    // { dg-warning "unused name" "" { target *-*-* } .-1 }
+}

